src/python/pants/rules/core/list_roots_test.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import json
+
+from pants.engine.fs import Digest, PathGlobs, Snapshot
+from pants.rules.core import list_roots
+from pants.source.source_root import SourceRoot, SourceRootCategories, SourceRootConfig
+from pants_test.console_rule_test_base import ConsoleRuleTestBase
+from pants_test.engine.util import run_rule
+from pants_test.test_base import TestBase
+
+
+class AllRootsTest(TestBase):
+  def test_all_roots(self):
+    SOURCE = SourceRootCategories.SOURCE
+    TEST = SourceRootCategories.TEST
+    THIRDPARTY = SourceRootCategories.THIRDPARTY
+
+    options = {
+      'pants_ignore': [],
+      'source_root_patterns': ['src/*', 'src/example/*'],
+      'source_roots': {
+        # Fixed roots should trump patterns which would detect contrib/go/examples/src/go here.
+        'contrib/go/examples/src/go/src': ['go'],
+
+        # Dir does not exist, should not be listed as a root.
+        'java': ['java']}
+    }
+    options.update(self.options[''])  # We need inherited values for pants_workdir etc.
+
+    self.context(for_subsystems=[SourceRootConfig], options={
```

Comment:
@benjyw : Do you see a way to make the `Subsystem` creation easier here?

Reply:
Not offhand, this is how we do it in v1 tests, no?

Reply:
Maybe. It's just... a lot.